src/vcpkg/binarycaching.cpp

```diff
@@ -409,37 +402,43 @@ namespace
         std::vector<path> m_write_dirs;
         std::vector<std::string> m_put_url_templates;
         std::vector<std::string> m_secrets;
-
-        std::set<PackageSpec> m_restored;
     };
-    struct HttpGetBinaryProvider : NullBinaryProvider
+    struct HttpGetBinaryProvider : IBinaryProvider
     {
         HttpGetBinaryProvider(std::vector<std::string>&& url_templates) : m_url_templates(std::move(url_templates)) { }
-        void prefetch(const VcpkgPaths& paths, std::vector<const Dependencies::InstallPlanAction*>& actions) override
+
+        RestoreResult try_restore(const VcpkgPaths&, const Dependencies::InstallPlanAction&) const override
         {
-            auto& fs = paths.get_filesystem();
+            return RestoreResult::unavailable;
+        }
 
-            const size_t current_restored = m_restored.size();
+        void push_success(const VcpkgPaths&, const Dependencies::InstallPlanAction&) const override { }
 
+        void prefetch(const VcpkgPaths& paths,
+                      const std::vector<Dependencies::InstallPlanAction>& actions,
+                      const std::vector<CacheStatus*>& cache_status) const override
+        {
+            auto& fs = paths.get_filesystem();
+            size_t this_restore_count = 0;
+            std::vector<std::pair<std::string, path>> url_paths;
+            std::vector<size_t> url_indices;
             for (auto&& url_template : m_url_templates)
             {
-                std::vector<std::pair<std::string, path>> url_paths;
-                std::vector<PackageSpec> specs;
-
-                for (auto&& action : actions)
+                url_paths.clear();
+                url_indices.clear();
+                for (size_t idx = 0; idx < actions.size(); ++idx)
                 {
-                    auto abi = action->package_abi();
-                    if (!abi)
+                    auto&& action = actions[idx];
+                    auto abi = action.package_abi();
+                    if (!abi || !cache_status[idx]->should_attempt_restore(this))
                     {
                         continue;
                     }
 
-                    specs.push_back(action->spec);
-                    auto pkgdir = paths.package_dir(action->spec);
-                    clean_prepare_dir(fs, pkgdir);
-                    pkgdir /= vcpkg::u8path(Strings::concat(*abi.get(), ".zip"));
-                    url_paths.emplace_back(Strings::replace_all(std::string(url_template), "<SHA>", *abi.get()),
-                                           pkgdir);
+                    clean_prepare_dir(fs, paths.package_dir(action.spec));
+                    url_paths.emplace_back(Strings::replace_all(url_template, "<SHA>", *abi.get()),
+                                           make_temp_archive_path(paths.buildtrees, action.spec));
```

Comment:
Highlighting that this looks like a behavior change: the zip is downloaded into `buildtrees\` instead of `packages\`.

Reply:
I think that's a good thing, after discussion made sure they all go into buildtrees